File: backend/app/ai/lesson.py
"""Lesson pipeline: outline -> scenes (OpenMAIC-inspired two-stage generation).

Stage 1 generates a structured lesson outline for a topic; stage 2 expands
each outline section into a typed scene on demand:

- ``explain`` / ``example`` scenes carry markdown narration plus an
  AI-classmate question (the question a confused learner would ask) with
  its answer.
- ``quiz`` scenes carry a multiple-choice check with an explanation.

The frontend walks scenes with prev/next playback and can export the
assembled lesson as a standalone HTML file.
"""
import re
import textwrap
from typing import Any, Dict, List, Optional

from ..logging_config import logger
from .engine import LocalLLMEngine
from .prompts import LEVEL_INSTRUCTIONS

SCENE_TYPES = ("explain", "example", "quiz")

OUTLINE_SYSTEM_PROMPT = textwrap.dedent("""\
    You design short math lessons. Produce a lesson outline as JSON only:

    {
      "title": "...",
      "sections": [
        {"title": "...", "type": "explain" | "example" | "quiz", "summary": "one sentence"},
        ...
      ]
    }

    Rules:
    - 4 to 6 sections.
    - Start with an "explain" section on the core idea.
    - Include at least one "example" (worked example) section.
    - End with exactly one "quiz" section.
    - Build from first principles: each section should depend only on the
      sections before it.
    """)

EXPLAIN_SYSTEM_PROMPT = textwrap.dedent("""\
    You write one scene of a math lesson. Respond with JSON only:

    {
      "narration": "markdown explanation of this section (150-300 words, use LaTeX \\\\(...\\\\) for math)",
      "classmate_question": "the question a curious but confused classmate would ask about this",
      "classmate_answer": "a friendly 2-3 sentence answer to that question"
    }

    Rules for narration:
    - Do NOT repeat the section title; start directly with the explanation.
    - Inline math: \\\\(...\\\\). Display math: $$...$$ on its own line.
    - Never use \\\\begin{equation} or other LaTeX environments.
    """)


def _strip_title_echo(narration: str, title: str) -> str:
    """Drop a leading line that just re-states the section title.

    Small models often open the narration with the title (sometimes with a
    trailing LaTeX "\\\\") even though the UI renders the title separately.
    """
    def _clean(s: str) -> str:
        s = re.sub(r"\\text(?:bf|it)\{([^}]*)\}", r"\1", s)
        for ch in "#*\\{}`":
            s = s.replace(ch, "")
        return " ".join(s.lower().split()).strip(" :.-")

    lines = narration.split("\n")
    i = 0
    while i < len(lines) and not lines[i].strip():
        i += 1
    if i < len(lines):
        a, b = _clean(lines[i]), _clean(title)
        # Only treat heading-sized lines as echoes — a one-line narration
        # that merely opens with the title must not be stripped away.
        heading_sized = len(a) <= max(len(b) * 2, len(b) + 30)
        if (
            a and b and heading_sized
            and (a == b or a.startswith(b) or b.startswith(a))
            and len(a) >= min(len(b) * 0.6, 12)
        ):
            return "\n".join(lines[i + 1:]).lstrip("\n")
        # Inline echo: "**Title:** rest of paragraph…" — drop just the prefix
        prefix = re.compile(
            r"^\s*(?:\\textbf\{|\*\*)\s*" + re.escape(title) + r"\s*(?:\}|\*\*)\s*[:.\-–—]*\s*",
            re.IGNORECASE,
        )
        if prefix.match(lines[i]):
            lines[i] = prefix.sub("", lines[i], count=1)
            return "\n".join(lines[i:])
    return narration

QUIZ_SYSTEM_PROMPT = textwrap.dedent("""\
    You write a quiz scene for a math lesson. Respond with JSON only:

    {
      "question": "one clear question testing the lesson's core idea",
      "choices": ["...", "...", "...", "..."],
      "correct_index": 0,
      "explanation": "why the correct answer is right, 1-2 sentences"
    }

    Rules: exactly 4 choices, plausible distractors, correct_index is the
    0-based index of the right answer.
    """)


class LessonService:
    def __init__(self) -> None:
        self._engine = LocalLLMEngine()

    def is_available(self) -> bool:
        return self._engine.is_available()

    # ------------------------------------------------------------------
    # Stage 1: outline
    # ------------------------------------------------------------------

    def outline(self, topic: str, level: str = "teen") -> Optional[Dict[str, Any]]:
        level_instruction = LEVEL_INSTRUCTIONS.get(level, "")
        raw = self._engine.chat_json(
            [
                {"role": "system", "content": OUTLINE_SYSTEM_PROMPT},
                {
                    "role": "user",
                    "content": f"{level_instruction}Design a lesson on: {topic[:300]}",
                },
            ],
            timeout_seconds=60,
            num_predict=800,
            temperature=0.4,
        )
        if not raw:
            return None
        sections = self._clean_sections(raw.get("sections"))
        if not sections:
            logger.warning("LessonService: outline had no usable sections")
            return None
        return {
            "title": str(raw.get("title") or topic),
            "level": level,
            "topic": topic,
            "sections": sections,
        }

    @staticmethod
    def _clean_sections(sections: Any) -> List[Dict[str, str]]:
        if not isinstance(sections, list):
            return []
        cleaned = []
        for s in sections[:8]:
            if not isinstance(s, dict) or not s.get("title"):
                continue
            stype = s.get("type") if s.get("type") in SCENE_TYPES else "explain"
            cleaned.append(
                {
                    "title": str(s["title"]),
                    "type": stype,
                    "summary": str(s.get("summary") or ""),
                }
            )
        return cleaned

    # ------------------------------------------------------------------
    # Stage 2: scene expansion
    # ------------------------------------------------------------------

    def scene(
        self,
        topic: str,
        level: str,
        section_title: str,
        section_type: str,
        summary: str = "",
    ) -> Optional[Dict[str, Any]]:
        if section_type == "quiz":
            return self._quiz_scene(topic, level, section_title, summary)
        return self._explain_scene(topic, level, section_title, section_type, summary)

    def _explain_scene(
        self, topic: str, level: str, title: str, stype: str, summary: str
    ) -> Optional[Dict[str, Any]]:
        level_instruction = LEVEL_INSTRUCTIONS.get(level, "")
        style = "Include a fully worked example with concrete numbers." if stype == "example" else ""
        from .library import get_library
        from .skills import COMPACT_SKILL

        library_context = get_library().context_for(f"{topic} {title}", k=2, max_chars=1200)
        messages = [
            {"role": "system", "content": EXPLAIN_SYSTEM_PROMPT + "\n" + COMPACT_SKILL},
            {
                "role": "user",
                "content": (
                    (library_context + "\n\n" if library_context else "")
                    + f"{level_instruction}Lesson topic: {topic[:200]}\n"
                    f"Section: {title}\n{summary}\n{style}"
                ),
            },
        ]
        raw = None
        narration = None
        # Small models occasionally truncate the JSON mid-narration or emit a
        # heading-only narration; retry once at lower temperature.
        for attempt in range(3):
            raw = self._engine.chat_json(
                messages,
                timeout_seconds=90,
                num_predict=1800,
                num_ctx=4096,
                temperature=0.4 if attempt == 0 else 0.6,
            )
            if not raw or not raw.get("narration"):
                continue
            candidate = _strip_title_echo(str(raw["narration"]), title)
            if len(candidate.strip()) >= 80:
                narration = candidate
                break
            logger.warning(
                f"LessonService: degenerate narration ({len(candidate.strip())} chars) "
                f"for '{title}', attempt {attempt + 1}"
            )
        if narration is None:
            return None
        return {
            "type": stype,
            "narration": narration,
            "classmate_question": str(raw.get("classmate_question") or ""),
            "classmate_answer": str(raw.get("classmate_answer") or ""),
        }

    def _quiz_scene(
        self, topic: str, level: str, title: str, summary: str
    ) -> Optional[Dict[str, Any]]:
        level_instruction = LEVEL_INSTRUCTIONS.get(level, "")
        messages = [
            {"role": "system", "content": QUIZ_SYSTEM_PROMPT},
            {
                "role": "user",
                "content": (
                    f"{level_instruction}Lesson topic: {topic[:200]}\n"
                    f"Quiz section: {title}\n{summary}"
                ),
            },
        ]
        raw = None
        for attempt in range(2):
            raw = self._engine.chat_json(
                messages,
                timeout_seconds=60,
                num_predict=800,
                num_ctx=4096,
                temperature=0.3 if attempt == 0 else 0.15,
            )
            if raw:
                break
        if not raw:
            return None
        choices = raw.get("choices")
        idx = raw.get("correct_index")
        if (
            not isinstance(choices, list)
            or len(choices) < 2
            or not isinstance(idx, int)
            or not 0 <= idx < len(choices)
            or not raw.get("question")
        ):
            logger.warning("LessonService: quiz scene malformed")
            return None
        return {
            "type": "quiz",
            "question": str(raw["question"]),
            "choices": [str(c) for c in choices[:4]],
            "correct_index": min(idx, 3),
            "explanation": str(raw.get("explanation") or ""),
        }

File: frontend/lesson.js
// AI Lessons — OpenMAIC-inspired outline → scene playback.
// Stage 1 fetches a lesson outline; stage 2 expands each section lazily.
// Scenes render with markdown-lite + KaTeX, an AI classmate persona asks
// a question per scene, quizzes are interactive, and the whole lesson can
// be exported as a standalone HTML file.
document.addEventListener("DOMContentLoaded", () => {
  const topicEl = document.getElementById("lesson-topic");
  const levelEl = document.getElementById("lesson-level");
  const startBtn = document.getElementById("lesson-start");
  const exportBtn = document.getElementById("lesson-export");
  const statusEl = document.getElementById("lesson-status");
  const bodyEl = document.getElementById("lesson-body");
  const titleEl = document.getElementById("lesson-title");
  const progressEl = document.getElementById("lesson-progress");
  const sceneEl = document.getElementById("lesson-scene");
  const prevBtn = document.getElementById("lesson-prev");
  const nextBtn = document.getElementById("lesson-next");
  if (!startBtn) return;

  let outline = null;
  let scenes = [];      // cached scene payloads by section index
  let current = 0;
  let historyStart = 0; // tutor history length when this lesson was built

  const askBtn = document.getElementById("lesson-ask");
  const askSceneBtn = document.getElementById("lesson-ask-scene");
  const tutorInput = document.getElementById("tutor-input");
  const tutorSolution = document.getElementById("tutor-solution");

  const TYPE_ICONS = { explain: "💡", example: "✏️", quiz: "❓" };

  function esc(s) {
    return String(s).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
  }

  // Minimal markdown: bold, italics, inline code, paragraphs, bullets
  function mdLite(text) {
    const lines = esc(text).split(/\n/);
    let html = "", inList = false;
    for (const line of lines) {
      const t = line.trim();
      if (/^[-*] /.test(t)) {
        if (!inList) { html += "<ul>"; inList = true; }
        html += "<li>" + t.slice(2) + "</li>";
      } else {
        if (inList) { html += "</ul>"; inList = false; }
        if (t) html += "<p>" + t + "</p>";
      }
    }
    if (inList) html += "</ul>";
    return html
      .replace(/\*\*([^*]+)\*\*/g, "<strong>$1</strong>")
      .replace(/\*([^*]+)\*/g, "<em>$1</em>")
      .replace(/`([^`]+)`/g, "<code>$1</code>")
      // LLMs sometimes emit LaTeX text commands outside math mode
      .replace(/\\textbf\{([^}]*)\}/g, "<strong>$1</strong>")
      .replace(/\\textit\{([^}]*)\}/g, "<em>$1</em>");
  }

  function renderMath(el) {
    if (typeof renderMathInElement === "function") {
      try {
        renderMathInElement(el, {
          delimiters: [
            { left: "$$", right: "$$", display: true },
            { left: "\\(", right: "\\)", display: false },
            { left: "\\[", right: "\\]", display: true },
            { left: "\\begin{equation}", right: "\\end{equation}", display: true },
            { left: "\\begin{align}", right: "\\end{align}", display: true },
          ],
          throwOnError: false,
          strict: "ignore",
        });
      } catch (e) { /* KaTeX optional */ }
    }
  }

  // LLM narration often re-states the section title as its own first line
  // (sometimes with a trailing LaTeX "\\") — drop it, the UI already shows
  // the title as a heading.
  function stripTitleEcho(text, title) {
    if (!title) return text;
    const clean = (s) =>
      String(s)
        .replace(/\\text(?:bf|it)\{([^}]*)\}/g, "$1")
        .toLowerCase()
        .replace(/[#*\\`{}]/g, "")
        .replace(/[\s:.\-]+$/g, "")
        .replace(/\s+/g, " ")
        .trim();
    const lines = String(text).split(/\n/);
    let i = 0;
    while (i < lines.length && !lines[i].trim()) i++;
    if (i < lines.length) {
      const a = clean(lines[i]);
      const b = clean(title);
      // Only heading-sized lines count as echoes; never strip a one-line
      // narration that merely opens with the title.
      const headingSized = a.length <= Math.max(b.length * 2, b.length + 30);
      if (a && b && headingSized && (a === b || a.startsWith(b) || b.startsWith(a)) &&
          a.length >= Math.min(b.length * 0.6, 12)) {
        lines.splice(0, i + 1);
        return lines.join("\n");
      }
      // Inline echo: "**Title:** rest of paragraph…" — drop just the prefix
      const escRe = title.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      const prefixRe = new RegExp(
        "^\\s*(?:\\\\textbf\\{|\\*\\*)\\s*" + escRe + "\\s*(?:\\}|\\*\\*)\\s*[:.\\-\\u2013\\u2014]*\\s*", "i");
      if (prefixRe.test(lines[i])) {
        lines[i] = lines[i].replace(prefixRe, "");
        return lines.join("\n");
      }
    }
    return text;
  }

  // Full narration → HTML: normalize LaTeX delimiters (shared helper from
  // app.js), turn equation-style environments into display math, and pull
  // display blocks out as placeholders so mdLite's per-line <p> wrapping
  // can't split them.
  function mdBlock(text, title) {
    let t = stripTitleEcho(String(text), title);
    if (typeof window.normalizeLatexDelimiters === "function") {
      t = window.normalizeLatexDelimiters(t);
    }
    // normalizeLatexDelimiters emits doubled delimiters (\\( … \\)) when it
    // converts $-math; collapse them so KaTeX's \( \) / \[ \] can match.
    t = t.replace(/\\\\([()[\]])/g, "\\$1");
    t = t.replace(
      /\\begin\{(equation|align|gather|displaymath)\*?\}([\s\S]*?)\\end\{\1\*?\}/g,
      (m, env, body) => "\n\\[" + body.trim() + "\\]\n"
    );
    const mathBlocks = [];
    t = t.replace(/\\\[([\s\S]*?)\\\]/g, (m, body) => {
      mathBlocks.push("\\[" + body.trim() + "\\]");
      return "\n%%MATH" + (mathBlocks.length - 1) + "%%\n";
    });
    let html = mdLite(t);
    mathBlocks.forEach((mb, i) => {
      const token = "%%MATH" + i + "%%";
      const block = '<div class="math-display" style="margin:10px 0;overflow-x:auto;">' + esc(mb) + "</div>";
      html = html.includes("<p>" + token + "</p>")
        ? html.replace("<p>" + token + "</p>", block)
        : html.replace(token, esc(mb));
    });
    return html;
  }

  function renderProgress() {
    progressEl.innerHTML = "";
    outline.sections.forEach((s, i) => {
      const chip = document.createElement("button");
      chip.className = i === current ? "btn-primary" : "btn-secondary";
      chip.style.cssText = "font-size:11px;padding:4px 10px;";
      chip.textContent = `${TYPE_ICONS[s.type] || ""} ${i + 1}. ${s.title}`;
      chip.addEventListener("click", () => showScene(i));
      progressEl.appendChild(chip);
    });
  }

  function classmateBlock(q, a) {
    if (!q) return "";
    return `
      <div style="margin-top:14px;padding:10px 12px;background:#f5f5f4;border-left:3px solid #a8a29e;border-radius:4px;">
        <div style="font-size:13px;"><strong>🧑‍🎓 Maya asks:</strong> ${mdBlock(q)}</div>
        <details style="margin-top:6px;font-size:13px;">
          <summary style="cursor:pointer;color:#57534e;">See the answer</summary>
          <div style="margin-top:6px;">${mdBlock(a)}</div>
        </details>
      </div>`;
  }

  function renderScene(idx) {
    const section = outline.sections[idx];
    const scene = scenes[idx];
    let html = `<div style="padding:6px 4px;">
      <div style="font-size:12px;color:#78716c;margin-bottom:6px;">
        Scene ${idx + 1} of ${outline.sections.length} • ${section.type}</div>
      <h4 style="margin:0 0 10px;">${esc(section.title)}</h4>`;

    if (scene.type === "quiz") {
      html += `<p style="font-weight:600;">${mdLite(scene.question)}</p><div id="quiz-choices">`;
      scene.choices.forEach((c, i) => {
        html += `<button class="btn-secondary quiz-choice" data-i="${i}"
          style="display:block;width:100%;text-align:left;margin:6px 0;padding:8px 12px;">${esc(c)}</button>`;
      });
      html += `</div><div id="quiz-feedback" style="margin-top:8px;font-size:13px;"></div>`;
    } else {
      html += mdBlock(scene.narration, section.title);
      html += classmateBlock(scene.classmate_question, scene.classmate_answer);
    }
    html += "</div>";
    sceneEl.innerHTML = html;
    renderMath(sceneEl);

    if (scene.type === "quiz") {
      sceneEl.querySelectorAll(".quiz-choice").forEach((btn) => {
        btn.addEventListener("click", () => {
          const i = parseInt(btn.dataset.i, 10);
          const fb = sceneEl.querySelector("#quiz-feedback");
          const right = i === scene.correct_index;
          btn.style.borderColor = right ? "#16a34a" : "#dc2626";
          fb.innerHTML = (right ? "✅ <strong>Correct!</strong> " : "❌ Not quite — try again, or ") +
            (right ? mdLite(scene.explanation) : "peek: <em>" +
              esc(scene.choices[scene.correct_index]) + "</em>");
        });
      });
    }
  }

  async function fetchScene(idx) {
    if (scenes[idx]) return scenes[idx];
    const s = outline.sections[idx];
    const resp = await fetch("/api/ai/lesson/scene", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        topic: outline.topic,
        level: outline.level,
        section_title: s.title,
        section_type: s.type,
        summary: s.summary || "",
      }),
    });
    if (!resp.ok) {
      const detail = (await resp.json().catch(() => ({}))).detail;
      throw new Error(detail || `HTTP ${resp.status}`);
    }
    scenes[idx] = await resp.json();
    return scenes[idx];
  }

  async function showScene(idx) {
    current = idx;
    window.EWTutor?.setLessonContext({
      topic: outline.topic,
      sceneIndex: idx,
      sceneTitle: outline.sections[idx].title,
    });
    renderProgress();
    prevBtn.disabled = idx === 0;
    nextBtn.disabled = idx >= outline.sections.length - 1;
    if (!scenes[idx]) {
      sceneEl.innerHTML = `<div style="padding:20px;color:#78716c;">✏️ Writing this scene with the local LLM…</div>`;
      statusEl.textContent = `Generating scene ${idx + 1}…`;
      try {
        await fetchScene(idx);
      } catch (err) {
        sceneEl.innerHTML = `<div style="padding:20px;color:#b91c1c;">Scene failed: ${esc(err.message)}
          <button class="btn-secondary" id="scene-retry" style="margin-left:8px;">Retry</button></div>`;
        sceneEl.querySelector("#scene-retry").addEventListener("click", () => showScene(idx));
        statusEl.textContent = "";
        return;
      }
      statusEl.textContent = "";
    }
    renderScene(idx);
    exportBtn.disabled = !scenes.some(Boolean);
  }

  startBtn.addEventListener("click", async () => {
    const topic = (topicEl.value || "").trim();
    if (!topic) { statusEl.textContent = "Enter a topic first."; return; }
    startBtn.disabled = true;
    statusEl.textContent = "Designing the lesson outline…";
    bodyEl.style.display = "none";
    outline = null; scenes = []; current = 0;
    try {
      const resp = await fetch("/api/ai/lesson/outline", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ topic, level: levelEl.value }),
      });
      if (!resp.ok) {
        const detail = (await resp.json().catch(() => ({}))).detail;
        throw new Error(detail || `HTTP ${resp.status}`);
      }
      outline = await resp.json();
      titleEl.textContent = outline.title;
      bodyEl.style.display = "";
      statusEl.textContent = "";
      // Scope the tutor's context store to this lesson
      await window.EWTutor?.startNewSession?.();
      historyStart = (window.EWTutor?.getHistory?.() || []).length;
      await showScene(0);
    } catch (err) {
      statusEl.textContent = "Lesson failed: " + err.message;
    } finally {
      startBtn.disabled = false;
    }
  });

  prevBtn.addEventListener("click", () => current > 0 && showScene(current - 1));
  nextBtn.addEventListener("click", () => current < outline.sections.length - 1 && showScene(current + 1));

  // "Just Ask" — free-form tutor question straight from the topic box,
  // no lesson required (and not scoped to any lesson).
  if (askBtn) {
    askBtn.addEventListener("click", () => {
      const q = (topicEl.value || "").trim();
      if (!q) { statusEl.textContent = "Type a question first."; return; }
      window.EWTutor?.setLessonContext(null);
      window.EWTutor?.ask(q);
      tutorSolution?.scrollIntoView({ behavior: "smooth", block: "start" });
    });
  }

  // "Ask about this scene" — focus the tutor input; app.js adds the
  // lesson/scene context to whatever gets asked.
  if (askSceneBtn) {
    askSceneBtn.addEventListener("click", () => {
      if (!tutorInput) return;
      if (!tutorInput.value.trim() && outline) {
        tutorInput.value = `Can you explain "${outline.sections[current].title}" in more depth?`;
      }
      tutorInput.focus();
      tutorInput.scrollIntoView({ behavior: "smooth", block: "center" });
    });
  }

  // ---------------------------------------------------------------
  // Export the assembled lesson as a standalone HTML file
  // ---------------------------------------------------------------
  exportBtn.addEventListener("click", () => {
    if (!outline) return;
    let content = "";
    outline.sections.forEach((s, i) => {
      const scene = scenes[i];
      content += `<section><h2>${TYPE_ICONS[s.type] || ""} ${esc(s.title)}</h2>`;
      if (!scene) {
        content += `<p><em>(scene not generated)</em></p>`;
      } else if (scene.type === "quiz") {
        content += `<p><strong>${esc(scene.question)}</strong></p><ol type="A">`;
        scene.choices.forEach((c, ci) => {
          content += `<li>${esc(c)}${ci === scene.correct_index ? " ✅" : ""}</li>`;
        });
        content += `</ol><p><em>${esc(scene.explanation)}</em></p>`;
      } else {
        content += mdBlock(scene.narration, s.title);
        if (scene.classmate_question) {
          content += `<blockquote><strong>🧑‍🎓 Maya asks:</strong> ${esc(scene.classmate_question)}<br/>
            <strong>Answer:</strong> ${esc(scene.classmate_answer)}</blockquote>`;
        }
      }
      content += "</section>";
    });
    // Include tutor Q&A asked while this lesson was open
    const history = (window.EWTutor?.getHistory?.() || []).slice(historyStart);
    if (history.length) {
      content += `<section><h2>💬 Questions asked during this lesson</h2>`;
      for (let i = 0; i < history.length; i++) {
        const m = history[i];
        if (m.role === "user") {
          content += `<p><strong>Q:</strong> ${esc(m.content)}</p>`;
        } else {
          content += `<blockquote>${mdBlock(m.content)}</blockquote>`;
        }
      }
      content += "</section>";
    }
    const doc = `<!DOCTYPE html><html><head><meta charset="utf-8">
<title>${esc(outline.title)}</title>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js"><\/script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js"
  onload="renderMathInElement(document.body,{delimiters:[{left:'\\\\(',right:'\\\\)',display:false},{left:'\\\\[',right:'\\\\]',display:true},{left:'$$',right:'$$',display:true}],throwOnError:false});"><\/script>
<style>
body{font-family:Georgia,serif;max-width:760px;margin:40px auto;padding:0 20px;color:#1c1917;line-height:1.6;}
h1{border-bottom:2px solid #1c1917;padding-bottom:8px;}
section{margin:28px 0;}
blockquote{background:#f5f5f4;border-left:3px solid #a8a29e;padding:10px 14px;border-radius:4px;}
footer{margin-top:40px;font-size:12px;color:#78716c;border-top:1px solid #e7e5e4;padding-top:10px;}
</style></head><body>
<h1>${esc(outline.title)}</h1>
<p><em>Level: ${esc(outline.level)} • Generated by Euclid's Window (local AI)</em></p>
${content}
<footer>Exported from Euclid's Window — learn math from first principles.</footer>
</body></html>`;
    const blob = new Blob([doc], { type: "text/html" });
    const a = document.createElement("a");
    a.href = URL.createObjectURL(blob);
    a.download = outline.title.replace(/[^a-z0-9]+/gi, "-").toLowerCase() + ".html";
    a.click();
    URL.revokeObjectURL(a.href);
  });
});
